clippy_lints/src/loops/infinite_loop.rs

```diff
@@ -128,3 +140,65 @@ impl<'hir> Visitor<'hir> for LoopVisitor<'hir, '_> {
         }
     }
 }
+
+/// Similar to [`FnRetTy`], but reveals the actual type of an `OpaqueDef`.
+enum RetTy<'hir> {
+    DefaultReturn(Span),
+    Return(Ty<'hir>),
+}
+
+impl<'hir> RetTy<'hir> {
+    fn from_fn_ret_ty(cx: &LateContext<'hir>, fn_ret_ty: FnRetTy<'hir>) -> Self {
+        /// Reveal and return the related type of an `opaque`, return `None` if the
+        /// given `ty` is not an `OpaqueDef`.
+        fn inner_<'tcx>(cx: &LateContext<'tcx>, ty: &Ty<'tcx>) -> Option<Ty<'tcx>> {
+            /// Visitor to find the type binding.
+            struct BindingVisitor<'tcx> {
+                res: Option<Ty<'tcx>>,
+            }
+            impl<'tcx> Visitor<'tcx> for BindingVisitor<'tcx> {
+                fn visit_assoc_type_binding(&mut self, type_binding: &'tcx hir::TypeBinding<'tcx>) {
+                    if self.res.is_some() {
+                        return;
+                    }
+                    if let hir::TypeBindingKind::Equality {
+                        term: hir::Term::Ty(ty),
+                    } = type_binding.kind
+                    {
+                        self.res = Some(*ty);
+                    }
+                }
+            }
+
+            let TyKind::OpaqueDef(item_id, ..) = ty.kind else {
+                return None;
+            };
+            let opaque_ty_item = cx.tcx.hir().item(item_id);
+
+            // Since `item_id` is from a `TyKind::OpaqueDef`,
+            // the `Item` related to it should always be `OpaqueTy`.
+            assert!(matches!(opaque_ty_item.kind, ItemKind::OpaqueTy(_)));
+
+            let mut vis = BindingVisitor { res: None };
+            vis.visit_item(opaque_ty_item);
+            vis.res
+        }
+
+        match fn_ret_ty {
+            FnRetTy::DefaultReturn(span) => Self::DefaultReturn(span),
+            FnRetTy::Return(ty) => Self::Return(inner_(cx, ty).unwrap_or(*ty)),
+        }
+    }
+    /// Returns the span to where the suggestion should be.
+    fn sugg_span(&self) -> Option<Span> {
+        match self {
+            Self::DefaultReturn(span) => Some(*span),
+            Self::Return(ty) if matches!(ty.kind, TyKind::Tup(&[])) => Some(ty.span),
+            Self::Return(_) => None,
+        }
+    }
+    fn is_never(&self) -> bool {
+        let Self::Return(ty) = self else { return false };
+        matches!(ty.kind, TyKind::Never)
```

Comment:
Could be `matches!(self, Self::Return(ty) if ty.is_never())`.

Reply:
ah, that's an `hir::Ty`, so there's no `is_never` calls, but I got what you mean, thanks